src/controllers/settings.js

```diff
@@ -2,16 +2,198 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import AppConstants from '../app-constants.js'
+
+import {
+  getUserEmails,
+  resetUnverifiedEmailAddress,
+  addSubscriberUnverifiedEmailHash,
+  removeOneSecondaryEmail,
+  getEmailById,
+  verifyEmailHash
+} from '../db/tables/email_addresses.js'
+
+import { setAllEmailsToPrimary } from '../db/tables/subscribers.js'
+
+import { fluentError, getMessage } from '../utils/fluent.js'
+import { sendEmail, getVerificationUrl, getUnsubscribeUrl } from '../utils/email.js'
+
+import { getBreachesForEmail } from '../utils/hibp.js'
+import { generateToken } from '../utils/csrf.js'
+
 import { mainLayout } from '../views/main.js'
 import { settings } from '../views/partials/settings.js'
+import { getTemplate } from '../emails/templates/email-2022.js'
+import { verifyPartial } from '../emails/partials/email-verify.js'
+
+async function settingsPage (req, res) {
+  const emails = await getUserEmails(req.session.user.id)
+  // Add primary subscriber email to the list
+  emails.push({
+    email: req.session.user.primary_email,
+    sha1: req.session.user.primary_sha1,
+    primary: true,
+    verified: true
+  })
+
+  const breachCounts = new Map()
+
+  const allBreaches = req.app.locals.breaches
+  for (const email of emails) {
+    const breaches = await getBreachesForEmail(
+      email.sha1,
+      allBreaches,
+      true,
+      false
+    )
+    breachCounts.set(email.email, breaches?.length || 0)
```

Comment:
Is this correct usage of optional chaining?  [From MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#description):

>Optional chaining cannot be used on a non-declared root object, but can be used with a root object with value undefined.

```js
undeclaredVar?.prop; // ReferenceError: undeclaredVar is not defined
```

Reply:
Hm, wouldn't `breaches` always be declared right above this? From a quick look at `getBreachesForEmail()` it looks like it might return `undefined`, so this would just avoid throwing in that case:

```js
> const test = undefined
> test?.length
undefined
> test.length
Uncaught TypeError: Cannot read properties of undefined (reading 'length')
```

Reply:
Yes, that makes sense and fits the exception given by MDN that I didn't read clearly!  Marking this resolved 👍